WIP: Add option to show guids on list command

cmungall opened the pull request:
Proposed: `-l` or `--long`, analogous to unix `ls`, for #118 

This is to show the `guid`, which is the 5 letter code after osf.io. Download URLs can be formed from `osf.io/{guid}/download`. This is useful to make a mapping between logical file names and download URLs

TBD:

 * what other properties to show?
 * allow other separator than `,`? Escape `,`s in filenames (because people)

BUGS:

sometimes the OSF API will return `None` for guids for files, even if they are there in the web interface. Need to check API docs more then contact support if required.

Note:

original proposal was to extend the upload command, but this doesn't seem to return guids and putting this on `list` seems more generally useful.

betatim commented on osfclient/cli.py:
```diff
@@ -224,7 +224,10 @@ def list_(args):
+            vals = [os.path.join(prefix, path)]
+            if args.long:
+                vals.append(str(file_.guid))
+            print(",".join(vals))
```
There is the `csv` module which might be handy to take care of all the escaping of characters and stuff so that we don't become one of those tools that produce broken CSV.

cmungall replied:
This one? https://pypi.python.org/pypi/csv

This seems a bit more fleshed out:
https://pypi.python.org/pypi/csvkit

any prefs? I only have experience of pandas which seems like a heavyweight dependency here.

betatim replied:
I was thinking of https://docs.python.org/3/library/csv.html

betatim commented:
To fix up the test take a look at the "fake args" mock we use. It is a bit basic in that we have to list all possible attributes by hand right now but seems to get the job done.

ctb commented:
The csv module in the Python stdlib --

https://docs.python.org/3/library/csv.html

cmungall commented:
OK, switched to using builtin csv.

Will add tests before requesting review

The main problem I am facing is that guid does not seem to be reliably returned by the OSF REST API. It seems you have to use the web interface and click on a file once for a guid to be granted.

For example, here is the results of one of my projects:

```
$ ./venv/bin/osf -p 2qk53 list -l
Please input your password: 
osfstorage/current/pre/mondo.owl,None
osfstorage/current/pre/mondo.obo,None
osfstorage/current/pre/mondo.json,None
osfstorage/current/README.md,None
osfstorage/current/mondo.owl,None
osfstorage/current/mondo.obo,None
osfstorage/current/mondo.json,None
osfstorage/pre/pre/mondo.owl,mhqcv
osfstorage/pre/pre/mondo.obo,nuvbs
osfstorage/pre/pre/mondo.json,4hj8b
```

I then went to the interface, and selected one of the files  with a None (`current/mondo.obo`), which took me here:

https://osf.io/e87hn/

next iteration, the guid appears:

```
$ ./venv/bin/osf -p 2qk53 list -l
Please input your password: 
osfstorage/current/pre/mondo.owl,None
osfstorage/current/pre/mondo.obo,None
osfstorage/current/pre/mondo.json,None
osfstorage/current/README.md,None
osfstorage/current/mondo.owl,None
osfstorage/current/mondo.obo,e87hn
osfstorage/current/mondo.json,None
osfstorage/pre/pre/mondo.owl,mhqcv
osfstorage/pre/pre/mondo.obo,nuvbs
osfstorage/pre/pre/mondo.json,4hj8b
```

This is highly suboptimal from my POV. I need to do more investigation to see if there is something in the existing API that can get around this, if not I will email their support people.

But won't get to this for a couple of weeks don't expect any action on this PR from me 'til then, sorry :(

betatim commented:
Does someone from the OSF know if the behaviour of not getting a GUID until after clicking on a file in the GUI is expected? Is there something we could do via the API to "generate" a GUID?

cmungall commented:
I just fired them an email asking this question

caseyrollins commented:
Hi @cmungall,

You are correct in that file GUIDs are not generated until the file is viewed, which would make building the shortened `osf.io/{guid}/download` URL a bit difficult. 

However, the API does return download URLs, even if a GUID does not yet exist. You can find this URL in the `links['download']` section of the response. The URL will have the form `https://files.osf.io/v1/resources/{node_id}/providers/osfstorage/{file_id}`. 

Hopefully this helps!

Casey

cmungall commented:
Thanks @caseyrollins 

I updated the PR, it now has 3 columns: path, guid and download. The guid is often None, as described above, but some users may like to see it where present. The download always appears to be present.

Output may look like this:

```
osfstorage/current/mondo.owl,bqpjm,https://files.osf.io/v1/resources/2qk53/providers/osfstorage/59e84168594d9002620e8b76
osfstorage/current/mondo.obo,e87hn,https://files.osf.io/v1/resources/2qk53/providers/osfstorage/59e833466c613b02612d3ef1
osfstorage/current/mondo.json,pbq3a,https://files.osf.io/v1/resources/2qk53/providers/osfstorage/59e832c2b83f690263734d8e
```

I note that these URLs have the version embedded. I think it's good practice to have this in APIs but what I'd like to avoid is a large scale remapping of all my URLs if one day v1 is deprecated and removed. I feel slightly more comfortable with the guids. 

Other disadvantages of the download URLs

 - they don't seem to be visible when browsing the files on the main website. Maybe I'm a bit obsessive this way but I like humans and agents to have access to the same kind of information
 - it's harder to find metadata about them. Given one of these URLs, I can easily parse the project, or I can download it, but if the contents don't provide much of a hint it's harder to find more about it. Contrast with guids, where I can just paste in the URL without the `/download/` part and see all versions etc.

I think the best we can do here is provide both in the `list` option, and let the user decide which to use for their use case, but I'm curious what others thoughts on this topic are.

ctb commented:
Random thoughts --

I think OSF does content-addressible storage, I wonder if the last
hash in the download URL is stable?

I don't believe the download URLs are guaranteed stable while I think the
guids are (vice changes in the file, etc.)

--t

cmungall commented:
Thanks for the thoughts Titus

Yes, it seems the guids are stable but not guaranteed for the download URLs. @caseyrollins can you confirm?

It may help to say something about my use cases.

I originally came to OSF as a potential solution for distributing bio-ontologies as part of the [Open Bio Ontologies](http://obofoundry.org) Library (see [this ticket](https://github.com/OBOFoundry/OBOFoundry.github.io/issues/494)). In this project we issue our own Permanent URLs (purl.obolibrary.org/obo/....), which also serve as GUIDs, and we map these to whatever delivery mechanism the ontology provider chooses [code here](https://github.com/OBOFoundry/purl.obolibrary.org/). For this use case, our users are insulated from any changes to the download URLs, though it would be a bit of a headache for our admins if we had to remap (e.g. if we used the OSF download URLs, and we were later forced to change from v1 to v2).

But as a part of the NIH Data Commons project I'm also interested in how OSF can play a role here (as I'm sure @ctb is too). One option is that DC assigns its own GUIDs and maps these to a variety of storage layers, similar to what OBO or DOI does but on a much larger scale. But it's better to avoid unnecessary layers, and OSF is most of the way there with its existing GUIDs, so if we can get things working smoothly here (i.e being able to get a GUID via the API rather than just the UI) then this might make a nice part of the commons ecosystem.

betatim commented:
@caseyrollins is there a way to use the OSF API to trigger the generation of a GUID? If yes then we could trigger this before printing out the list.

brianjgeiger commented:
So we are somewhat protective of our 5-character GUID namespace. The major purpose for guids is to ensure that someone has a really simple way to refer to their resource in a space-constrained medium such as a print publication or telling someone the URL over the phone. We have gone through great efforts to keep people from programmatically generating guids for large numbers of items at a time and I don't see that changing anytime soon.

That being said, any URLs that you can access through osf.io (as opposed to api.osf.io) are guaranteed to reach the resource that they point to, presuming that the resource is still public and hasn't been deleted. We do not guarantee query parameters, but everything before.

For the case at hand, I can definitely see your concerns. The URLs we are providing aren't meant for long-term persistence. I think I see a way that we can fix that with relatively little effort without polluting our GUID namespace, but I'd like to chat with some folks this morning and I'll let you know what we come up with.

brianjgeiger commented:
The plan is to create persistent routes for downloads. When we have them (probably in January seems the most likely timeframe), they'll replace out the existing download links and will likely be in the form `osf.io/<node_id>/download/<file-id-or-guid/`. This won't break existing code, so the API won't need to be versioned.

aubreymoore commented:
Here's another use case which requires download links for files:

I am working on a project to extract label data from about 5,000 images of herbarium specimens stored in OSF Storage. Part of the project will involve crowdsourcing, so I will need download links for the images. I have not figured out how to find download links using the API. Is there a way?

brianjgeiger commented:
@aubreymoore Check out https://api.osf.io/v2/files/553e69248c5e4a219919ea54/ data->links->download for an example

cmungall commented:
Thanks @brianjgeiger!

OK, so I suggest as far as the CLI goes we can go ahead and have the list behavior show both guid and existing download link, possibly with a note about expectations for each linking back to this PR. We can update the note when the new download links are rolled out. Users of the cli can choose to use whichever URL best suits their purposes however they choose.

I still have to go and make the mock tests which at the current rate I'm going might be another week or so, sorry!

aubreymoore commented:
Thanks @cmungall 

Is the proposed improvement to the list command going to be released soon? I am publishing a methods article and I'd really like to refer to osfclient in my published workflow as the way to get download links from OSF.

cmungall commented:
I'll likely not get to the mock tests this year..

felliott commented:
Apologies, @cmungall, I prematurely merged this into a `develop` branch that I ended up rebasing away.  I ran into problems getting the `csv` module working on py2.7.   As of `v0.0.5` osfclient is no longer prioritizing py2.7 support, so this can be merged without trouble.  @jh27539 added some unit tests, and I've merged them all into current `develop` (pre-v0.0.6).  They will go up on Pypi when `v0.0.6` is released.  Thank you!

Cheers,
@felliott
